opentreemap/works_management/models.py

```diff
@@ -21,10 +22,28 @@ class Team(models.Model):
 class WorkOrder(Auditable, models.Model):
     instance = models.ForeignKey(Instance)
     name = models.CharField(max_length=255, null=False, blank=False)
+    reference_num = models.IntegerField(default=0)
     created_at = models.DateTimeField(auto_now_add=True)
     created_by = models.ForeignKey(User)
     updated_at = models.DateTimeField(auto_now=True)
 
+    class Meta:
+        unique_together = ('instance', 'reference_num')
+
+    def get_next_reference_num(self):
+        """
+        Return next sequential reference number for instance.
+        """
+        agg_result = Instance.objects.filter(id=self.instance.id) \
+            .annotate(max_value=Max('workorder__reference_num'))[0]
+        max_value = agg_result.max_value or 0
+        return max_value + 1
```

Comment:
In general, it's a good idea to avoid the practice of requesting a value from the database, manipulating it, and saving it to the database.  It leads to race conditions where multiple workers may see the same max value.

A way to avoid the race condition is to get and manipulate the maximum number in the database, which will handle concurrency issues for us. Django has support for this with [Query Expressions](https://docs.djangoproject.com/en/1.8/ref/models/expressions/). Another would be to lock the `Task` / `WorkOrder` table during the transaction using [`select_for_update()`](https://docs.djangoproject.com/en/1.10/ref/models/querysets/#select-for-update).